pyadjoint/overloaded_type.py

```diff
@@ -142,6 +142,19 @@ def _ad_restore_at_checkpoint(self, checkpoint):
         """
         raise NotImplementedError
 
+    def _ad_assign(self, other):
+        """This method must be overridden.
+
+        The method should implement a routine for assigning other to `self`.
+
+        Note:
+            This method should not be employed for immutable types.
```

Comment:
I think this line here highlights an issue with this approach. `OverloadedType` actually conflates two sets of functionality:
1. Things needed in order to tape a variable.
2. Things needed in order to work with variables when computing taped operations (replay, tlm, adjoint etc.) and optimisation.

This functionality is part of 2 (it's not needed for taping). At that stage of operations, all variables are treated as immutable, so assignment looks like a strange operation at best. The fact that it doesn't work for some types is a further red flag.

Assignment itself is also not really what is meant here. The use case in question seems to be assigning a scalar to a vector, which is not well-defined (think about non-point-evaluation FunctionSpaces, for example). I think what is really meant here is interpolate. 

On the mutability point, I think this should be more like `_ad_init_object` and return a new value. I.e. the signature would be:
```
    @classmethod
    def _ad_interpolate(cls, obj, value):
```
and the required semantics would be to return a new object which is identical to `obj` except that its value approximates `value`.

Reply:
`_ad_iadd` and `_ad_imul` are existing in-place operations. However I agree here an in-place operation is not needed. I think the easiest is just to change `VecInterface.to_petsc` to handle the scalar case using `VecSetValues`.

I also don't think scalar bounds are clearly defined. These could mean

- Bounds the dof values. I think this is in the intention, and is what `VecSetValues` would do.
- Bound function values. Too difficult, but see #145.
- Bound interpolated function values, but this is backend specific. Interpolation may not be meaningfully defined.

Reply:
This would bring us back to the option of requiring the user to specify bounds of the control type. I think this is probably the right thing -  Pyadjoint can't interpret the meaning of variables beyond basic Hilbert space operations (i.e. vector space operations + inner product). 

I guess the corner case is the no bounds case. We could either make that a user problem or we'd have to add `_ad` methods for producing variables that are element-wise minimal or maximal. That would be convenient for users but I'm not sure it smells good.

Reply:
> requiring the user to specify bounds of the control type

I'd definitely prefer that. Existing pyadjoint examples use scalars, though.

> I guess the corner case is the no bounds case.

That at least can be handled with `VecSetValues` and `np.finfo(PETSc.ScalarType).min`/`np.finfo(PETSc.ScalarType).max`.

Reply:
I've implemented the 'scalar bound = bound on all dofs' case using 'VecSet`'. This might still lead to some surprises depending on the basis, but I think is consistent with pyadjoint behaviour elsewhere.

Reply:
Current pyadjoint behaviour is described here.

https://github.com/dolfin-adjoint/pyadjoint/blob/92121af97711be88ff6d7994c0c7469129b2fb25/pyadjoint/optimization/optimization_problem.py#L24-L28

Reply:
The actual check allows `None` and `int` as well.
https://github.com/dolfin-adjoint/pyadjoint/blob/92121af97711be88ff6d7994c0c7469129b2fb25/pyadjoint/optimization/optimization_problem.py#L51-L55

Reply:
As discussed in the meeting, this should be fixed separately from this PR.